ambry-commons/src/main/java/com.github.ambry.commons/HelixNotifier.java

```diff
@@ -0,0 +1,175 @@
+      throw new IllegalArgumentException("path to listen cannot be null");
+    }
+    if (topicListener == null) {
+      throw new IllegalArgumentException("listener cannot be null");
+    }
+    String topicPath = makeTopicPath(topic);
+    HelixPropertyListener helixListener = new HelixPropertyListener() {
+      @Override
+      public void onDataChange(String path) {
+        logger.trace("Message is changed for topic {} at path {}", topic, path);
+        sendMessageToLocalTopicListener(topicListener, topic, path);
+      }
+
+      @Override
+      public void onDataCreate(String path) {
+        logger.trace("Message is created for topic {} at path {}", topic, path);
+        sendMessageToLocalTopicListener(topicListener, topic, path);
+      }
+
+      @Override
+      public void onDataDelete(String path) {
+        // for now, this is a no-op when a ZNRecord for a topic is deleted, since no
+        // topic deletion is currently supported.
+        logger.debug("Message is deleted for topic {} at path {}", topic, path);
+      }
+    };
+    topicListenerToHelixListenerMap.put(topicListener, helixListener);
+    helixStore.subscribe(topicPath, helixListener);
+    logger.trace("TopicListener={} has been subscribed to topic={}", topicListener, topic);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void unsubscribe(String topic, TopicListener<String> topicListener) {
+    if (topic == null) {
+      throw new IllegalArgumentException("topic cannot be null");
+    }
+    if (topicListener == null) {
+      throw new IllegalArgumentException("topicListener cannot be null");
+    }
+    String topicPath = makeTopicPath(topic);
+    HelixPropertyListener helixListener = topicListenerToHelixListenerMap.remove(topicListener);
+    helixStore.unsubscribe(topicPath, helixListener);
+    logger.trace("TopicListener={} has been unsubscribed from topic={}", topicListener, topic);
+  }
+
+  /**
+   * Sends a message to local {@link TopicListener} that subscribed topics through this {@code HelixNotifier}.
+   * @param topicListener The {@link TopicListener} to send the message.
+   * @param topic The topic which the message belongs to.
+   * @param path The path to the topic.
+   */
+  private void sendMessageToLocalTopicListener(TopicListener topicListener, String topic, String path) {
+    String message = null;
+    try {
+      ZNRecord zNRecord = helixStore.get(path, null, AccessOption.PERSISTENT);
+      if (zNRecord != null) {
+        message = zNRecord.getSimpleField(MESSAGE_KEY);
+        long startTimeMs = System.currentTimeMillis();
+        topicListener.onMessageReceive(topic, message);
+        logger.trace("Message has been sent to TopicListener={} on topic={} with message={} using {}ms", topicListener,
+            topic, message, System.currentTimeMillis() - startTimeMs);
+      } else {
+        logger.debug("TopicListener={} on topic={} should receive a message, but the corresponding ZNRecord is null",
```

Comment:
Normally this would not happen. There is a slight chance that before getting the `ZNRecord`, the node is deleted (for now there is no such deletion operation). To not encounter `NullPointerException`, just make a check.